### PR TITLE
Add tests and docs for environment variable resolution.

### LINKED ISSUE #3100: Formalize rules that govern environment variable inheritance and precedence
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/issues/3173

Originally created by @pnadolny13 on 2022-01-20 19:40:37

---

I've had a couple use cases where I'd like to set different environment variables at different levels. For sensitive credentials I set them in the .env while non-sensitive environment specific settings (like schema_name) get set at the meltano.yml environment level.

The functionality I want is to be able to combine both sensitive and non-sensitive variables for use in a plugin as an environment variable. A use case is in the squared repo where we have an Athena connection string used in Great Expectations and Superset:

`awsathena+rest://[YOUR_AWS_KEY]:[YOUR_AWS_SECRET]@athena.us-east-2.amazonaws.com/[YOUR_SCHEMA_NAME]?s3_staging_dir=s3://[YOUR_S3_PATH]`

I want to be able to set the secret keys in my .env and the path and schema in my meltano.yml environment. Right now templating is supported in meltano.yml for config values using environment variables but for some plugins, especially utilities, configs arent useful and we need to be building environment variables that the utility can access. Templating doesnt seem to be supported at the `env` level underneath the `environment` key. So right now I can set secrets in the .env and non-sensitive variables at the `environment.env` level in meltano.yml but I have no way of blending the two to pass along to a plugin, so I end up building the connection string multiple times in the plugins vs defining it once.

Ideas: Add an `env` key to plugin configs that allows you to use templating to build a new environment variable at the plugin/environment level. Or...allow templating at the current `env` key level based on whats in your .env file.

## Update (2022-05-05)

Scope of this issue is now to:

* Write tests based on the spec in the [Miro Board](https://miro.com/app/board/uXjVO3j_ADw=/)
* Fix anything that fails tests
* Update documentation to clarify inheritance

## Update (2022-05-03)

**Relevant [Miro Board](https://miro.com/app/board/uXjVO3j_ADw=/) where we're whiteboarding all of this**

## Spec Update (2022-04-08)

Note: this updated spec follows from my comment here: https://gitlab.com/meltano/meltano/-/issues/3173#note_817409586

### Problem statement:

Currently, env variables at different levels of the code cannot necessarily rely on each other or build upon each other.

### Proposed solution:

We would define environment variable inheritance to be valid across these layers, in this strict order:

- Level 1: terminal context (inherited either via `.env` or via shell env vars)
- Level 2: top-level `env:` declarations within `meltano.yml`
- Level 3: environment-level `env:` declarations.
- Level 4: `env:` declarations within top-level plugin definitions.
- Level 5: `env:` declarations within environment-scoped plugin definitions.

### Example

This shows how inheritance would be expected to function across the 5 named levels. (Could be a unit test during development.)

`.env`

```sh
# Level 1: terminal context
# Inherits from none
LEVEL_NUM="1"
STACKED="1"
```

```yml
env:
  # Level 2: top-level `env:`
  # Inherits from terminal context
  LEVEL_NUM: "2"                  #  '2'
  STACKED: "${STACKED}2"          # '12'
plugins:
  extractors:
    tap-foobar:
      env:
        # Level 4: plugin-level `env:`
        # Inherits from a environment-level `env:` if an environment is active
        # Inherits directly from top-level `env:` if no environment is active
        LEVEL_NUM: "4"            #    '4'
        STACKED: "${STACKED}4"    # '1234'
environments:
  prod:
    env:
      # Level 3: environment-level `env:`
      # Inherits from top-level `env:`
      LEVEL_NUM: "3"              #   '3'
      STACKED: "${STACKED}3"      # '123'
  plugins:
    extractors:
      tap-foobar:
        env:
          # Level 5: environment-level plugin `env:`
          # Inherits from (global) plugin-level `env:`
          LEVEL_NUM: "5"          #     '5'
          STACKED: "${STACKED}5"  # '12345'
```

## FAQ

**Q. Why does the global plugin level config (level 4) inherit from the environment-level config?**

A: This is counterintuitive at first but actually quite necessary.

Imagining that the plugin settings globally include a call to `$SNOWFLAKE_USERNAME` and imagine that each environment has its own value for `SNOWFLAKE_USERNAME: ___` in the respective `env:`. We don't want each environment to have to redundantly declare the plugin properties but we _do_ want to use the `$SNOWFLAKE_USERNAME` value that is most appropriate from the given execution context. Hence, the global-level plugin can be defined once to leverage whatever value is in `$SNOWFLAKE_USERNAME` and the plugin's configuration will still be seeded properly according to which environment is active.

## Providing level info for research/debugging

Perhaps in the `dry-run` proposal in [#3146](https://gitlab.com/meltano/meltano/-/issues/3146), it would be good to print where each environment variables are sourced. We will need to do this without exposing sensitive values.

So, if the example above were real, we might get debug info such as:

```
meltano run --dry tap-foobar target-snowflake
Printing debug info in response to `--dry` CLI argument. No actions will be executed.

The plugin `tap-foobar` was declared with the following context:

- env:LEVEL_NUM - derived from the following, in order of precedence:
    - environments[prod].plugins[tap-foobar].env[LEVEL_NUM]
    - plugins[tap-foobar].env[LEVEL_NUM]
    - environments[prod].env[LEVEL_NUM]
    - env[LEVEL_NUM]
    - shell.env[LEVEL_NUM]
...
```

Or perhaps a more realistic example:

```
meltano run --dry tap-foobar target-snowflake
Printing debug info in response to `--dry` CLI argument. No actions will be executed.

The plugin `tap-foobar` was declared with the following context:

- 'tap-foobar.config[username]' was derived from the following, in order of precedence:
    - environments[prod].env[FOOBAR_USERNAME]
    - env[FOOBAR_USERNAME]
...
```

In the above example, the 'prod' declaration of `FOOBAR_USERNAME` clobbers the global one. The user can infer from this printout that their `prod`-level config is working correctly, and if they were to remove the prod-level config, they'd get the value of the top-level `env:` declaration.

## Related Topics

More discussion on this can and probably should be moved to a follow on issue, but for completeness:

### Further detail and settings-level `env:` declarations (`level 6`)

Not listed above, but a plugin setting may have `env:` declared, which operates differently from the other layers. This `plugin.settings[].env:` declaration simply accepts a string as env variable name, and the specified setting value will get injected and sent _into_ the plugin's subprocess. 

- This functionality is required in order to have plugins behave as expected. Non-singer plugins don't receive a `config.json` so the env variable injection is the substitute.
- I think the `env:` function at the setting layer should _always_ be a pure child process injection (clobbering any existing env var but not reading from it).
- We should probably stop this layer 6 from _also_ sourcing data from layers 1-5, and instead make it a pure one-way injection into the plugin context. As described in [#3447](https://gitlab.com/meltano/meltano/-/issues/3447)
- I do not know if the current status quo for an existing `env` variable value will clobber an explicitly defined config value, or if the config value would clobber the env variable value, essentially replacing the env var value within the plugin's child process.

### PR DESCRIPTION
**Pull request by MeltyBot**
Migrated from GitLab: https://gitlab.com/meltano/meltano/-/merge_requests/2667

Originally created by @cjohnhanson on 2022-05-26 21:59:55

---

## Merge Request Checklist

- [X] Link to an issue
- [X] Include the proposed fix or feature
- [X] Include and update tests for the modified code
- [X] Include a documentation change
- [X] Add a [CHANGELOG.md](https://gitlab.com/meltano/meltano/blob/master/README.md#Changelog) entry in the Unreleased section for any user-facing changes
- [X] Mention or assign a maintainer

Closes [#3173](https://gitlab.com/meltano/meltano/-/issues/3173)

### CODE REVIEW

**Review by tayloramurphy — CHANGES_REQUESTED** [review]
@aaronsteers @cjohnhanson I'm realizing that the configuration page https://docs.meltano.com/guide/configuration doesn't mention the `env:` key _at all_... 

@cjohnhanson I think for this PR we need something in the `Environment variables` section  that mentions that key. 

@aaronsteers so as not to completely blow up scope on the PR, I can open an additional issue around the env key since we currently have that info scattered all around.

**Comment by tayloramurphy** [comment]
@aaronsteers I opened https://github.com/meltano/meltano/issues/5969

File: docs/src/_reference/command-line-interface.md [hunk 1]
```diff
@@ -591,6 +591,20 @@ To execute plugins inside containers, use the `--containers` flag:
...
+When debugging plugin configuration, it is often useful to view environment variables being provided to a plugin at runtime.
+This can be achieved using `--log-level=debug` but for readability and convenience, the `meltano invoke` command also supports printing individual environment variables to stdout at runtime:
+```bash
+# Print the runtime value PLUGIN_ENVIRONMENT_VARIABLE_1:
+meltano invoke --print-environment-variable PLUGIN_ENVIRONMENT_VARIABLE_1
+
+# The option supports printing multiple variables as well.
+
+# # Print the runtime values of both PLUGIN_ENVIRONMENT_VARIABLE_1 and PLUGIN_ENVIRONMENT_2 on separate lines:
+meltano invoke --print-environment-variable PLUGIN_ENVIRONMENT_VARIABLE_1 --print-environment-variable PLUGIN_ENVIRONMENT_VARIABLE_2
```
[the remark this answers is not in the document]
Comment (cjohnhanson) [reply docs/src/_reference/command-line-interface.md]: Yep they definitely are, I added a placeholder plugin for these examples.

**Comment by cjohnhanson** [comment]
@aaronsteers - 
> Do you mind adding summary of tests/changes here, and especially if there are any additional tasks or issues to resolve?

Sure thing, the high-level takeaway is that there are 7 tests which currently `xfail` (rather than hard fail, which is why you won't see these failures in the pipeline, they'll show as `XFAIL`s in pytest output). These 7 tests can be resolved by addressing two underlying causes.

Shell environment variables take too high of precedence compared to what is described in the miro board, I'll create a follow-up issue for that which will resolve 6 of the currently `xfail'-ing tests. 

Inheritance (e.g., setting a `STACKED` env var to `1` and then `{STACKED}2`, etc.)  also basically doesn't work at all as expected. I'll create another follow-up issue to address that which will resolve the one remaining failing test case. 

To be clear though, the behavior _is_ in line with what our docs _currently_ say, but not with the new model as described in the miro board.

> and/or are there any other caveats that should be called out before merging this PR and closing the related issue?
I still need to address this feedback from @tayloramurphy :
> @cjohnhanson I think for this PR we need something in the Environment variables section that mentions that key.

See also this comment in the gitlab MR for some more details:
> The tests I've written are in line with my reading of the "New: Simplified Model" for precedence in [the miro board](https://miro.com/app/board/uXjVO3j_ADw=/), but differ significantly from current behavior. The tests corresponding to the scenarios in which current behavior is out of line with desired behavior are currently set to `xfail` and are parametrized with human-readable scenario descriptions to facilitate this follow-up work. The current behavior for precedence matches what's currently in [the docs](https://docs.meltano.com/guide/configuration#configuration-layers), though I added a short clarification on environment-level precedence for plugin settings.The tests I've written are in line with my reading of the "New: Simplified Model" for precedence in [the miro board](https://miro.com/app/board/uXjVO3j_ADw=/), but differ significantly from current behavior. The tests corresponding to the scenarios in which current behavior is out of line with desired behavior are currently set to xfail and are parametrized with human-readable scenario descriptions to facilitate this follow-up work. The current behavior for precedence matches what's currently in [the docs](https://docs.meltano.com/guide/configuration#configuration-layers), though I added a short clarification on environment-level precedence for plugin settings.

**Comment by cjohnhanson** [comment]
@aaronsteers @tayloramurphy I'm hoping to get to a PR tonight for my one other remaining 2.0 issue. So if I misunderstood scope here and we need to try to address one or both of the resolution/inheritance issues I outlined in my above comment, I can push to get that taken care of in the next couple days, or potentially stretching into Monday of next week.

**Comment by aaronsteers** [comment]
@cjohnhanson - Re: 

> These 7 tests can be resolved by addressing two underlying causes.
> 
> Shell environment variables take too high of precedence compared to what is described in the miro board, I'll create a follow-up issue for that which will resolve 6 of the currently `xfail'-ing tests.
> 
> Inheritance (e.g., setting a STACKED env var to 1 and then {STACKED}2, etc.) also basically doesn't work at all as expected. I'll create another follow-up issue to address that which will resolve the one remaining failing test case.

Thanks very much for this additional detail. I really like the approach of using xfail tests so we can keep those in the codebase and work those towards resolution.

> So if I misunderstood scope here and we need to try to address one or both...

Actually, I think you've got the scope perfectly. If I understand correctly, there are basically three todos I'd ask before closing:

1. [ ] Create new issues for those 2 follow-on fixes as described above (and link them in a comment here for ref).
2. [ ] Can we add inline refs for each respective xfail tests, linking in a code comment to the logged issue that is expected to fix the failing test.
3. [ ] Not covered above, but I see failing tests still in pytest suite on GitLab. Is there additional work needed to get these to pass? https://gitlab.com/meltano/legacy-ci/meltano/-/jobs/2536112067

I think with these three items resolved, we can close. What I can't speak to yet is whether we'll move to prioritize either of the two new issues. I'll wait to review those more closely before I and @tayloramurphy try to make a call on relative priority. (If you feel like you can add a `weight` estimate on each, that could be very helpful.)

**Comment by aaronsteers** [comment]
```
>       assert len(result_no_ff.settings) == 11  # noqa: WPS432
E       assert 12 == 11
```

**Comment by cjohnhanson** [comment]
@aaronsteers there was some additional work needed to get these to pass -- I had introduced a regression in a recent commit.

Those specific failing tests are passing locally now. I'll check in on the pipeline later today and get those issues created and then add the inline refs next to the xfails, and then this should be good to merge+close today.

**Review by edgarrmondragon — APPROVED** [review]

**Comment by cjohnhanson** [comment]
@tayloramurphy is this what you had in mind for a quick blurb about `env:` key?
https://github.com/meltano/meltano/pull/5938/files#diff-627749b70e154d1698d041017804e255927fac4cac7383dd6ee3fd6fd05a3b70L128

**Comment by tayloramurphy** [comment]
> @tayloramurphy is this what you had in mind for a quick blurb about `env:` key? https://github.com/meltano/meltano/pull/5938/files#diff-627749b70e154d1698d041017804e255927fac4cac7383dd6ee3fd6fd05a3b70L128

@cjohnhanson yep! That looks great - I have #5969 to track adding more 👍

**Review by tayloramurphy — APPROVED** [review]

**Comment by aaronsteers** [comment]
@cjohnhanson - re:

> is this what you had in mind for a quick blurb about env: key?
> https://github.com/meltano/meltano/pull/5938/files#diff-627749b70e154d1698d041017804e255927fac4cac7383dd6ee3fd6fd05a3b70L128


Love this! Thanks!

**Pull request merged**